commons/src/scanner/symbol/iterator/matcher.rs

```diff
@@ -0,0 +1,150 @@
+/// Function type to consume prefix sequences of a new line.
+pub type IteratorPrefixFn = Rc<dyn (Fn(&mut dyn PrefixMatcher) -> bool)>;
+
+/// Trait containing functions that are available inside the end matcher function.
+pub trait EndMatcher {
+    /// Returns `true` if the upcoming [`Symbol`] sequence is an empty line.
+    /// Meaning that a line contains no [`Symbol`] or only [`SymbolKind::Whitespace`].
+    ///
+    /// **Note:** This is also `true` if a parent iterator stripped non-whitespace symbols, and the nested iterator only has whitespace symbols.
+    ///
+    /// [`Symbol`]: super::Symbol
+    fn is_empty_line(&mut self) -> bool;
+
+    /// Wrapper around [`Self::is_empty_line()`] that additionally consumes the matched empty line.
+    /// Consuming means the related iterator advances over the matched empty line.
+    ///
+    /// **Note:** The iterator is only advanced if an empty line is matched.
+    ///
+    /// **Note:** The empty line is **not** included in the symbols returned by [`SymbolIterator::take_to_end()`].
+    fn consumed_is_empty_line(&mut self) -> bool;
+
+    /// Returns `true` if the given [`Symbol`] sequence matches the upcoming one.
+    ///
+    /// [`Symbol`]: super::Symbol
+    fn matches(&mut self, sequence: &[SymbolKind]) -> bool;
+
+    /// Wrapper around [`Self::matches()`] that additionally consumes the matched sequence.
+    /// Consuming means the related iterator advances over the matched sequence.
+    ///
+    /// **Note:** The iterator is only advanced if the sequence is matched.
+    ///
+    /// **Note:** The matched sequence is **not** included in the symbols returned by [`SymbolIterator::take_to_end()`].
+    fn consumed_matches(&mut self, sequence: &[SymbolKind]) -> bool;
+
+    /// Returns `true` if the iterator is at the given nesting depth.
+    ///
+    /// **Note** Use [`SymbolIterator::curr_depth()`] to get the current depth of an iterator.
+    fn at_depth(&self, depth: usize) -> bool;
+}
+
+/// Trait containing functions that are available inside the prefix matcher function.
+pub trait PrefixMatcher {
+    /// Consumes and returns `true` if the given [`Symbol`] sequence matches the upcoming one.
+    /// Consuming means the related iterator advances over the matched sequence.
+    ///
+    /// **Note:** The iterator is only advanced if the sequence is matched.
+    ///
+    /// **Note:** The given sequence must **not** include any [`SymbolKind::Newline`], because matches are only considered per line.
+    ///
+    /// **Note:** The matched sequence is **not** included in the symbols returned by [`SymbolIterator::take_to_end()`].
+    ///
+    /// [`Symbol`]: super::Symbol
+    fn consumed_prefix(&mut self, sequence: &[SymbolKind]) -> bool;
+}
+
+impl<'input> EndMatcher for SymbolIterator<'input> {
+    fn is_empty_line(&mut self) -> bool {
+        // Note: Multiple matches may be set in the match closure, so we need to ensure that all start at the same index
```

Comment:
I did not know that about `NOTE`.
But would you also write `NOTE` in doc-comments?
Usually I write `**Note:**` in doc-comments, but because normal comments have no formatting, I stayed with `Note` here, because it felt more consistent.

Would have to look through the code so replace all `Note`. Maybe keeping it as is for now, and change all in a new PR?

Reply:
Hmm good question. Generally I wouldn't write any form of `Note:` in doc-comments. Would probably just explain it, something like `Note that ...`.

You can decide if and when you want to change this, it's not important part of this PR anyway.

Reply:
I think notes in doc-comments can be useful.
Think of them more like [GitHub alerts](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts).

It helps to highlight information that is especially relevant to a user.
Making a section bold does not make it readable, so you use alerts to highlight those sections instead.

Reply:
I know, but just in general keep in mind that it's not necessary most of the time. If something is that important, maybe separate heading is a better option. Otherwise we can just explain it. I also use `NOTE:` often, but we should probably reserve it for special cases. It kind of loses it's purpose if we over-use it, just want us to be aware of that.